src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -172,11 +172,25 @@ internal BoundExpression ConvertPatternExpression(TypeSymbol inputType, CSharpSy
 
             // If we are pattern-matching against an open type, we do not convert the constant to the type of the input.
             // This permits us to match a value of type `IComparable<T>` with a pattern of type `int`.
-            if (inputType.ContainsTypeParameter() &&
-                // But do not permit matching null against a struct type.
-                (expression.ConstantValue != ConstantValue.Null || !inputType.IsNonNullableValueType()))
+            bool inputContainsTypeParameter = inputType.ContainsTypeParameter();
+            HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+            if (inputContainsTypeParameter)
```

Comment:
Why does it matter if the input contains a type parameter? Shouldn't we report the same warning for `new C() is 3`?